expression/builtin_encryption.go

```diff
@@ -112,9 +130,42 @@ func (b *builtinAesDecryptSig) evalString(row chunk.Row) (string, bool, error) {
 		return "", true, errors.Trace(err)
 	}
 
-	// TODO: Support other modes.
-	key := encrypt.DeriveKeyMySQL([]byte(keyStr), aes128ecbBlobkSize)
-	plainText, err := encrypt.AESDecryptWithECB([]byte(cryptStr), key)
+	var iv string
+	modeName, _ := b.ctx.GetSessionVars().GetSystemVar(variable.BlockEncryptionMode)
+	mode, exists := aesModes[strings.ToLower(modeName)]
+	if !exists {
+		return "", true, errors.Errorf("unsupported block encryption mode - %v", modeName)
+	}
+	if mode.ivRequired {
+		if len(b.args) != 3 {
+			return "", true, ErrIncorrectParameterCount.GenByArgs("aes_decrypt")
+		}
+		iv, isNull, err = b.args[2].EvalString(b.ctx, row)
+		if isNull || err != nil {
+			return "", true, errors.Trace(err)
+		}
+		if len(iv) < aes.BlockSize {
+			return "", true, errIncorrectArgs.Gen("The initialization vector supplied to aes_decrypt is too short. Must be at least 16 bytes long")
+		}
+		// init_vector must be 16 bytes or longer (bytes in excess of 16 are ignored)
+		iv = iv[0:aes.BlockSize]
+	} else {
+		if len(b.args) == 3 {
+			// For modes that do not require init_vector, it is ignored and a warning is generated if it is specified.
+			b.ctx.GetSessionVars().StmtCtx.AppendWarning(errWarnOptionIgnored.GenByArgs("IV"))
+		}
+	}
+
+	key := encrypt.DeriveKeyMySQL([]byte(keyStr), mode.keySize)
+	var plainText []byte
+	switch mode.modeName {
+	case "ecb":
+		plainText, err = encrypt.AESDecryptWithECB([]byte(cryptStr), key)
+	case "cbc":
+		plainText, err = encrypt.AESDecryptWithCBC([]byte(cryptStr), key, []byte(iv))
```

Comment:
It seems some duplicate logic in `AESDecryptWithECB` and `AESDecryptWithCBC`, `AESEncryptWithECB` and `AESEncryptWithCBC`~

what about use `AESDecrypt`, `AESEncrypt`, then inversion of control, pass a `cipher.BlockMode` argument?

Reply:
I add "type blockModeBuild func(block cipher.Block) cipher.BlockMode" to create BlockMode argument.